### PR TITLE
implemented support magnet URI extension, select specific file indices for download, BEP53

### PR DESCRIPTION
**Pull request by aldenml**

### CODE REVIEW

File: test/test_magnet.cpp [hunk 1]
```diff
@@ -476,3 +476,82 @@ TORRENT_TEST(invalid_web_seed_escaping)
...
 
+TORRENT_TEST(parse_magnet_select_only)
+{
+	error_code ec;
+	add_torrent_params p = parse_magnet_uri("magnet:?xt=urn:btih:cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd"
+		"&dn=foo&so=0,2,4,6-8", ec);
+	TEST_CHECK(!ec);
+
+	TEST_EQUAL(p.file_priorities.size(), 9);
+	for (std::size_t i = 0; i < p.file_priorities.size(); i++)
```
Comment (arvidn) [inline test/test_magnet.cpp]: I think this would be clearer by doing:
```
auto const yes = default_priority;
auto const no = dont_download;
TEST_CHECK(p.file_priorities == std::vector<download_priority_t>{
   yes, no, yes, no, yes, no, yes, yes, yes});
```
what do you think?

File: test/test_magnet.cpp [hunk 2]
```diff
@@ -476,3 +476,82 @@ TORRENT_TEST(invalid_web_seed_escaping)
...
+		{
+			TEST_EQUAL(p.file_priorities[i], default_priority);
+		}
+		else
+		{
+			TEST_EQUAL(p.file_priorities[i], dont_download);
+		}
+	}
+
+	p = parse_magnet_uri("magnet:?xt=urn:btih:cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd"
```
Comment (arvidn) [inline test/test_magnet.cpp]: if it's not too much trouble, I think these should be separated into individual test cases. If something breaks, it's a lot easier to pin-point which case broke in that case

File: test/test_magnet.cpp [hunk 3]
```diff
@@ -476,3 +476,82 @@ TORRENT_TEST(invalid_web_seed_escaping)
...
```
Comment (arvidn) [inline test/test_magnet.cpp]: does the BEP defined multiple arguments being additive?

File: ChangeLog [hunk 4]
```diff
@@ -1,3 +1,5 @@
+	* implemented support magnet URI extension, select specific file indices
+	  for download, BEP53
```
Comment (arvidn) [inline ChangeLog]: would you mind adding a bullet for this BEP on the ``docs/features.rst`` page as well? ideally with a link to bittorrent.org

File: src/magnet_uri.cpp [hunk 5]
```diff
@@ -258,14 +258,68 @@ namespace libtorrent {
...
 
+		auto select_pos = std::string::npos;
+		string_view select = url_has_argument(uri, "so", &select_pos);
+		while (!select.empty())
+		{
+			// parse the ranges or indices
+			do
+			{
+				string_view token;
+				std::tie(token, select) = split_string(select, ',');
```
Comment (arvidn) [inline src/magnet_uri.cpp]: technically this would accept quotes. say: ``&so="1,2"`` would not be interpreted the way you would expect, but also not fail. i.e. ``split_string`` doesn't split quoted substrings.

Comment (arvidn) [reply src/magnet_uri.cpp]: maybe this isn't the end of the world, but i think there should be some negative tests problably, of invalid input.

File: src/magnet_uri.cpp [hunk 6]
```diff
@@ -258,14 +258,68 @@ namespace libtorrent {
...
+
+				int idx1, idx2;
+				// TODO: what's the right number here?
+				constexpr int max_index = 10000; // can't risk out of memory
+
+				auto const divider = token.find_first_of('-');
```
Comment (arvidn) [inline src/magnet_uri.cpp]: I would imagine ``split_string()`` is an even better fit here, since you only expect there to be a single "-", so you get both sides back right away.

Comment (arvidn) [reply src/magnet_uri.cpp]: to know whether it was found or not, see if the second result is an empty string

File: src/magnet_uri.cpp [hunk 7]
```diff
@@ -258,14 +258,68 @@ namespace libtorrent {
...
+				if (divider != std::string::npos) // it's a range
+				{
+					if (divider == 0) // no start index
```
Comment (arvidn) [inline src/magnet_uri.cpp]: ideally there would be a test case for every failure that's handled here

Comment (arvidn) [reply src/magnet_uri.cpp]: including indices > 10000

**Comment by arvidn** [comment]
thanks for this!

File: ChangeLog (hunk 4 above)
Comment (aldenml) [reply ChangeLog]: OK

File: src/magnet_uri.cpp (hunk 6 above)
Comment (aldenml) [reply src/magnet_uri.cpp]: OK

File: src/magnet_uri.cpp (hunk 5 above)
Comment (aldenml) [reply src/magnet_uri.cpp]: I'm not sure I follow you here, could you rephrase please?

File: src/magnet_uri.cpp (hunk 7 above)
Comment (aldenml) [reply src/magnet_uri.cpp]: OK

File: test/test_magnet.cpp (hunk 1 above)
Comment (aldenml) [reply test/test_magnet.cpp]: yes, better :)

File: test/test_magnet.cpp (hunk 2 above)
Comment (aldenml) [reply test/test_magnet.cpp]: OK

File: test/test_magnet.cpp (hunk 3 above)
Comment (aldenml) [reply test/test_magnet.cpp]: no, that was my interpretation, it could be a reset, you tell me

File: src/magnet_uri.cpp (hunk 5 above)
Comment (arvidn) [reply src/magnet_uri.cpp]: if you look at the test for ``split_string()``, you'll see that it supports quotes. i.e. it won't split quoted strings

File: test/test_magnet.cpp (hunk 3 above)
Comment (arvidn) [reply test/test_magnet.cpp]: I don't have an opinion. perhaps it makes sense to bring it back to bittorrent.org and see if anyone else does

Comment (aldenml) [reply test/test_magnet.cpp]: @dessalines what is your position here?

File: src/magnet_uri.cpp (hunk 5 above)
Comment (aldenml) [reply src/magnet_uri.cpp]: I will add some protection here

File: src/magnet_uri.cpp (hunk 6 above)
Comment (aldenml) [reply src/magnet_uri.cpp]: looking at this again, I need the index to know if the `-` is at the beginning or right at the end

**Comment by Col-blimp** [comment]
`add_torrent_params` has `url` depreciated so there is no way to tell if a magnet link is a BEP53 link or not once it hits the session.

**Comment by arvidn** [comment]
@Col-blimp is that a question or a statement?
in principle that's right, it's definitely the direction of the new ``read_resume_data()`` and ``write_resume_data()``. However, as long as you build with deprecated functions enabled, that ``url`` field is still there, and should still be set to the original magnet link when you get the ``add_torrent_alert``. But maybe you had a different case in mind.

**Comment by Col-blimp** [comment]
It was pointing out a potential problem, BEP53 is like me deciding you take your coffee black when you want it with milk and sugar so it should be optional and there is nothing in the API to give you the choice, `url ` from  `add_torrent_params` would do the job as you could just look for `&so=` in the string but as it's depreciated you can't do that, BEP53 is not expected behaviour so I'm thinking you should be able to know if a link is or is not BEP53.

**Comment by arvidn** [comment]
ah, right. but with the non-deprecated behavior, you get to see the file priorities in ``add_torrent_params`` and you can alter them any way you'd like before adding it to the session

**Comment by Col-blimp** [comment]
non-deprecated is fine but enabling depreciated features to use a new feature doesn't seem right, put it another way, once a BEP53 link has been added to the session given `add_torrent_params` `url` is unavailable and `torrent_handle` `file_priorities()` is unavailable how can you on the `metadata_received_alert` determine if this is or is not a BEP53 torrent?

**Comment by aldenml** [comment]
I fail to see the nature of the problem (but I'm sure I'm missing something in your reasoning). I see this `&so=` similar to any other optional parameter like `&x.pe=`, if you discard the actual `add_torrent_params`, you will never know if peers connected were provided in the magnet.

I expect the UI can hold some information if relevant, also `add_torrent_alert` provide you with a copy of the original `add_torrent_params` (or at least I think it should)

**Comment by Col-blimp** [comment]
The problem was that the magnet link was removed from the `add_torrent_params` as was `file_priorities()` from `torrent_handle` so I could not figure out how to ignore this BEP without adding an external list of magnet links, `get_file_priorities()` has been added in place of `file_priorities()` so that can be gone through to see if anything is not `default_priority`.

**Comment by aldenml** [comment]
if you want to ignore it, just `(add_torrent_params)::file_priorities.clear()` should do the trick right after parsing any magnet.

**Comment by arvidn** [comment]
it's still not clear to me exactly what you want to achieve that might require an external list of magnet links.

was it just the querying priorities of files? (if so, referring to it as a "bep53-torrent" is a bit misleading, since file priorities exists independently of this feature). Even with the ``get_file_priorities()``, you can't tell whether those priorities originated from a magnet link or not. In fact, once the metadata has been received, I don't believe you can tell that the torrent originated from a magnet link either.

**Comment by Col-blimp** [comment]
Yes, I just wanted to know if the file priorities had been set by this BEP at the point of getting the `metadata_received_alert` so I could then set the corresponding UI dialog according to user choice.
